tests/test_class_formatting.rs

```diff
@@ -0,0 +1,159 @@
+#![cfg(feature = "macros")]
+
+use pyo3::prelude::*;
+use pyo3::py_run;
+use std::fmt::{Display, Formatter};
+
+#[path = "../src/tests/common.rs"]
+mod common;
+
+#[pyclass(eq, str = "MyEnum.{:?}")]
+#[derive(Debug, PartialEq)]
+pub enum MyEnum {
+    Variant,
+    OtherVariant,
+}
+
+#[pyclass(eq, str)]
+#[derive(Debug, PartialEq)]
+pub enum MyEnum2 {
+    Variant,
+    OtherVariant,
+}
+
+impl Display for MyEnum2 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[pyclass(eq, str = "MyEnum.{}")]
+#[derive(Debug, PartialEq)]
+pub enum MyEnum3 {
+    #[pyo3(name = "AwesomeVariant")]
+    Variant,
+    OtherVariant,
+}
```

Comment:
Hmm, similar to complex enum I feel like we should either allow specifying the display string per variant, or only delegate to `Display`. This feels a bit to magic for my taste, given that this is meant for simple cases a convenience and more complex cases can still implement `__str__` manually.

Also: Shouldn't this require `Display for MyEnum3` the way it is currently written?

Reply:
I agree that this case comes off as a bit of magic.  It is a nesting of convenience methods.  But this is exactly what happens in the current repr case for simple enums.  So much flexibility was given that it obscures intent when just reading the code.

The reason this doesn't implement Display directly is also a bit odd.  Basically, renaming builds a match case to map the variants to a String representing the name on the Python side.  Without the renaming you would write `#[pyclass(str="MyEnum.{:?}")` to accomplish the same result as the above referenced code that contains renaming.  That is because you are no longer passing a variant to the format macro, you are passing a string.  If you retained the debug format string, you would get `MyEnum."AwesomeVariant"`.  Changing the format string gives the output without the quotations in the output string.

I struggled with the behavior around these renaming cases.  On the Python side, the user will see these other names, and if the format string outputs the rust names in Python, this would be un-intuitive.  However.  supporting the renaming leads to these un-intuitive format string cases.  It feels like the high degree of flexiblity leads to an un-intuitive result one way or another.

Reply:
Ah, I was already wondering what this renaming business was about. (Haven't looked too closely at the implementation yet).

> If you retained the debug format string, you would get `MyEnum."AwesomeVariant"`.

Especially something like this is fragile and prone to errors with weird messages if you get it wrong. I don't think this is something that we should support or recommend.

We could allow `#[pyo3(str = "...")]` on the variants to do this, but I'm not sure if it is worth to add so much complexity here. Maybe a viable option is to add another attribute (`str_variant`?, not sure) that replicates the current behavior and have `str` only work on `Display`. These options would be exclusive, so at most one of the could be active. And a manual implementation of `__str__` would of course also be possible if neither attribute was given.

Reply:
@Icxolu I have made a check for the rename cases and made the string shorthand format incompatible with renaming.  This strange formatting scenario can no longer arise to surprise users.